### ISSUE TITLE
Deprecating base point compression and explicit curve parameters in EC key files

### ISSUE THREAD

**Issue by akiratk0355** [openssl/openssl#9286]
Hi all,

We are researchers in cryptography, and have recently published an attack paper targeting OpenSSL which we would like to draw your attention to:

A. Takahashi and M. Tibouchi. _Degenerate Fault Attacks on Elliptic Curve Parameters in OpenSSL._ [[pdf](https://eprint.iacr.org/2019/400.pdf)]
 
The attacks described in our paper are no cause for panic--they are probably only of immediate concern to a tiny fraction of OpenSSL users, running the software on embedded platforms and relying on highly unusual parameter choices. However, we believe that they illustrate the need to
deprecate certain exotic features of OpenSSL's implementation of elliptic curve crypto.

## Summary

Fault analysis is a type of physical attacks that seek to learn secrets by deliberately tampering with the device to cause malfunction or otherwise unexpected behaviour, by modifying the voltage of the power source at carefully chosen points in time, subjecting the device to sudden changes of temperature, etc.

We discovered that it is feasible to extract the secret key from a single faulty ECDSA signature if a suitably synchronized fault is injected during the signing operation, and experimentally verified that the attacks are reliably reproducible using OpenSSL 1.1.1 installed on a Raspberry Pi single board computer.

## Vulnerable EC key files and OpenSSL versions

EC key files subject to our attacks can be constructed with the following options to the ecparam command.

* the `-param_enc explicit` option, leading to explicit curve parameters,
  instead of OID for a named curve.

* the `-conv_form compressed` option, leading to a compressed base point
  and public key.
 
In short, the following command yields a vulnerable EC key file:

```
$ openssl ecparam -out eckey.pem -name <OID> -genkey -conv_form compressed -param_enc explicit
```

We confirmed that the above operation is possible in OpenSSL version 1.1.0h and 1.1.1 onwards.

## How the attacks work in a nutshell

Upon loading such vulnerable EC key files, by making use of an instruction skipping fault injection technique, our attacks can force the base points to be incorrectly decompressed into points on special "degenerate" curves, called singular and supersingular curves, where solving the discrete logarithm problem is computationally much easier.

The attack is especially devastating against curves with j-invariant equal to 0 such as the Bitcoin curve secp256k1, for which key recovery reduces to a single division in the base field.

## Is this a vulnerability of OpenSSL?

No. According to the OpenSSL security policy, physical fault injection is outside the scope of its threat model, and therefore these attacks are not classified as vulnerabilities.

## What are our suggestions?

Immediate action is only really needed on the part of users of OpenSSL who run it on devices subject to active physical attacks (such as IoT devices) and who, in addition, rely on curve parameters with compressed base points (so that users of named curves are not affected).
Realistically speaking, this is most likely a long-winded way of saying "practically nobody".

Nevertheless, we believe that our attacks illustrate how the following two features of the OpenSSL implementation of ECC can become serious pitfalls in terms of security, while offering essentially no benefit in return:

1. Allowing curve parameters with compressed base points: apart from a    tiny gain in terms of storage space (which only exists when using    custom curve parameters to begin with), there is no reason why anyone should use this, and many reasons to think this is a way for users to    shoot themselves in the foot.

2. Allowing custom elliptic curve parameters at all: although this is  allowed by RFC 4492, later RFCs superseding that one only include  support for named curves, presumably because custom curves are hard to implement securely, slow, and highly impractical to validate from a security standpoint.

For these reasons, we suggest deprecating 1 and preferably 2 as well.

## Further reading

Detailed write-up about this attack by Joël Alwen.

  https://wickr.com/new-attacks-on-openssls-elliptic-curve-crypto/

Slides from our presentation at EuroS&P'19 conference in Stockholm.

  https://akiratk0355.github.io/file/slides_EuroSP.pdf

We are of course happy to provide further details upon request.

Kind regards,

-- 
Akira Takahashi <takahashi@cs.au.dk> @akiratk0355 and
Mehdi Tibouchi <mehdi.tibouchi@normalesup.org> @mti

**Comment by slontis** [comment]
@romen

**Comment by levitte** [comment]
Ping @romen

**Comment by romen** [comment]
I agree, in an ideal world, we probably should aim for removing support for curves defined by explicit parameters: they are messy, ensuring security is complicated, they have been deprecated in TLS 1.2 (RFC 8422), they have seen little mainstream adoption anyway.

That being said, it is really up to the OMC to decide if the above ones are reasons enough to break backward compatibility and remove a feature, and as @akiratk0355 mentioned, physical fault attacks are outside the scope of the threat model established by the OpenSSL security policy, so most likely this particular issue cannot be considered as a security fix but more of a preemptive security feature.

As a somewhat relevant data point BoringSSL got rid of explicit parameter curves and they still manage to talk over the Internet. This says little about applications that have been built against OpenSSL and might be relying on explicit parameter curves, but we should consider if we are doing those users a service by continuing to support a security foot-cannon without warnings when alternatives that definitely have better performance, better safe-usability and better interoperability are to be preferred.

This should probably trigger a discussion in the @openssl/omc to understand how to move forward.

**Comment by romen** [comment]
# Measures that are _complementary_ to deprecation

This should not stop the discussion about deprecation, but I think it would be a good idea, in any case to have `#ifdef` guards triggerable using the build time configuration profiles to give users the option to compile OpenSSL with disabled support for `explicit curve parameters` and `compressed base points in explicit curve parameters`.

I believe this could be a good addition as it can allow to bring safer and saner behavior also to the users of currently supported branches.

**Comment by mattcaswell** [comment]
Everything @romen just said makes sense to me. We would not backport such changes in any case so we're only talking about what would we do for 3.0 and beyond. Deprecation seems a reasonable policy for these features since it does not actually result in immediate removal anyway - so we're not going to suddenly break applications relying on these features - but at the same time we give fair warning of some future removal.

Adding `#ifdef` guards does also make sense in this case - although I am slightly wary of adding lots of these. They end up being quite a high maintenance burden (I frequently have to go and fix "no-ec" and other similar options). OTOH they do offer users a way to customise the build and improve security where we cannot do this in the default build for whatever reason - so it is probably justifiable here.

PR welcome ;-)

**Comment by kroeckx** [comment]
You mention 1.1.0h and 1.1.1. In 1.1.0l and 1.1.1d when we read a
file with explicit parameters for a known curve, we've switched to
using the built-in curve. Does that mean that the example you give
doesn't actually work anymore with the latest version, but that
since that version you actually need to use a custom curve to be
vulnerable?

**Comment by romen** [comment]
> You mention 1.1.0h and 1.1.1. In 1.1.0l and 1.1.1d when we read a file with explicit parameters for a known curve, we've switched to using the built-in curve. Does that mean that the example you give doesn't actually work anymore with the latest version, but that since that version you actually need to use a custom curve to be vulnerable?

Good point @kroeckx I planned to mention it in my first comment and forgot.

The security implications reported by @akiratk0355 now should only be relevant for custom curves, not for the built-in curves, as we try to match explicit parameters against known curves when parsing such keys, and only if they are unknown we use them as custom curves.

**Comment by romen** [comment]
I just self unassigned to leave room for @akiratk0355 or any other willing contributor that wants to propose a PR for this issue!

**Comment by davidben** [comment]
> As a somewhat relevant data point BoringSSL got rid of explicit parameter curves and they still manage to talk over the Internet. This says little about applications that have been built against OpenSSL and might be relying on explicit parameter curves, but we should consider if we are doing those users a service by continuing to support a security foot-cannon without warnings when alternatives that definitely have better performance, better safe-usability and better interoperability are to be preferred.

Explicit parameter curves are indeed terrible. I strongly support getting rid of them. To expand on our BoringSSL's behavior somewhat:

* We do not support explicit parameter encoding in public keys (mostly of what one would get over the internet) at all. That change was made four years ago and we've not had reports of problems. When I was testing this, I believe the Firefox, macOS, and Windows parsers all rejected these, so that's further data points. RFC 5480 also [prohibits the explicit form](https://tools.ietf.org/html/rfc5480#section-2.1.1).
* We do support explicit parameter encoding of known curves in private keys (more commonly from configuration than the internet). This was done more out of caution than a concrete problem. I haven't gotten to trying to get rid of them. Note, though, OpenSSL 1.0.x defaulted to explicit encoding in some contexts. This parser recognizes the built-in curves but rejects all other inputs. We never parse explicit form into a custom curve.
* We do, sadly, support creating custom curves via the public API (`EC_GROUP_new_curve_GFp` and friends). This is purely for the benefit of one consumer (Conscrypt in Android), which needs it to implement some unfortunate Java EC APIs. Custom curves never match built-in curves for equality purposes (`EC_GROUP_cmp`), even if they happen to have the same parameters.
* Serializing or deserializing a custom curves in BoringSSL errors. We never emit explicit curve form, and we never decode anything to custom curve.

**Comment by akiratk0355** [comment]
> The security implications reported by @akiratk0355 now should only be relevant for custom curves, not for the built-in curves, as we try to match explicit parameters against known curves when parsing such keys, and only if they are unknown we use them as custom curves.

@kroeckx @romen  Thanks for the comments. 
I quickly checked with 1.1.1d by commenting out the vulnerable branch in `BN_bin2bn()`, and it looks like the same attack is still feasible against explicit known curve parameters.
I guess the reason why it still works is because that known curve check is executed after the domain parameter initialization, while our attack targets much earlier stage: [the initialization of the curve parameter "b"](https://github.com/openssl/openssl/blob/1d320e5c4cb90f75565465b89e6e5fd09cf2093b/crypto/ec/ec_asn1.c#L604). Hence the victim ends up having incorrect domain parameters after the fault, which of course doesn't match any built-in curves and will be used as is.

**Comment by paulidale** [comment]
This isn't going to happen for 3.0.  It is a good candidate for post 3.0.

**Comment by cryptechsurpport** [comment]
I started trading Crypto on Binary options, at first everything was going smoothly till I tried withdrawing. I was made to believe that I had to build my account to a certain point before withdrawing, I had lost total sum of 10.000btc. After weeks of turmoil and sadness that investment brought me and my family. I saw recommendations to 
P E R F E C T R E C L A I M R E C O V E R I N G O N   I N S T A and i was lucky they helped me recover all of my funds without any hassle in less than 48 hours, I am so happy doing this because I know how many people would benefit from this useful information. Fortunately for me I was able to recover all i lost P E R F E C T R E C L A I M RECOVERY CONSULTANT is a real refund professional feel free to reach out to them for quick response if you’re having similar issues

**Comment by cryptechsurpport** [comment]
I started trading Crypto on Binary options, at first everything was going smoothly till I tried withdrawing. I was made to believe that I had to build my account to a certain point before withdrawing, I had lost total sum of 10.000btc. After weeks of turmoil and sadness that investment brought me and my family. I saw recommendations to 
P E R F E C T R E C L A I M R E C O V E R I N G O N   I N S T A and i was lucky they helped me recover all of my funds without any hassle in less than 48 hours, I am so happy doing this because I know how many people would benefit from this useful information. Fortunately for me I was able to recover all i lost P E R F E C T R E C L A I M RECOVERY CONSULTANT is a real refund professional feel free to reach out to them for quick response if you’re having similar issues

**Comment by cryptechsurpport** [comment]
I started trading Crypto on Binary options, at first everything was going smoothly till I tried withdrawing. I was made to believe that I had to build my account to a certain point before withdrawing, I had lost total sum of 10.000btc. After weeks of turmoil and sadness that investment brought me and my family. I saw recommendations to 
P E R F E C T R E C L A I M R E C O V E R I N G O N   I N S T A and i was lucky they helped me recover all of my funds without any hassle in less than 48 hours, I am so happy doing this because I know how many people would benefit from this useful information. Fortunately for me I was able to recover all i lost P E R F E C T R E C L A I M RECOVERY CONSULTANT is a real refund professional feel free to reach out to them for quick response if you’re having similar issues

**Comment by cryptechsurpport** [comment]
I started trading Crypto on Binary options, at first everything was going smoothly till I tried withdrawing. I was made to believe that I had to build my account to a certain point before withdrawing, I had lost total sum of 10.000btc. After weeks of turmoil and sadness that investment brought me and my family. I saw recommendations to 
P E R F E C T R E C L A I M R E C O V E R I N G O N   I N S T A and i was lucky they helped me recover all of my funds without any hassle in less than 48 hours, I am so happy doing this because I know how many people would benefit from this useful information. Fortunately for me I was able to recover all i lost P E R F E C T R E C L A I M RECOVERY CONSULTANT is a real refund professional feel free to reach out to them for quick response if you’re having similar issues

**Comment by cryptechsurpport** [comment]
I started trading Crypto on Binary options, at first everything was going smoothly till I tried withdrawing. I was made to believe that I had to build my account to a certain point before withdrawing, I had lost total sum of 10.000btc. After weeks of turmoil and sadness that investment brought me and my family. I saw recommendations to 
P E R F E C T R E C L A I M R E C O V E R I N G O N   I N S T A and i was lucky they helped me recover all of my funds without any hassle in less than 48 hours, I am so happy doing this because I know how many people would benefit from this useful information. Fortunately for me I was able to recover all i lost P E R F E C T R E C L A I M RECOVERY CONSULTANT is a real refund professional feel free to reach out to them for quick response if you’re having similar issues

**Comment by t8m** [comment]
Explicit curves are now disabled. I do not think we want to disable point compression support for now.
